Add ckProfiler support for forward 3D convolutions with OUT element-wise operations.

andriy-ca opened the pull request:
Added `grouped_conv_fwd_outelementop` operation to ckProfiler.

The option enables performance profiling of 3D FWD convolutions on tensors with non-standard floating-point data types followed by scaling operation.

At this time, the following combinations of data types and operations can be profiled:
 - fp8@fp8->fp8 with ConvScale and ConvInvScale operations;
 - bf8@bf8->fp8 with ConvScale operation;
 - fp8@bf8->fp8 with ConvScale operation.

Support for profiling the following combinations is implemented, but CK currently does not instantiate corresponding instances:
- bf8@bf8->fp8 with ConvInvScale operation;
- fp8@bf8->fp8 with ConvInvScale operation;
- bf8@fp8->fp8 with ConvScale and ConvInvScale operations.

Refer to [grouped_convolution_forward_convscale.hpp](https://github.com/ROCm/composable_kernel/blob/6474b52da74b21bba0de916820afe92e2d9e579b/library/include/ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_convscale.hpp) and [grouped_convolution_forward_convinvscale.hpp](https://github.com/ROCm/composable_kernel/blob/6474b52da74b21bba0de916820afe92e2d9e579b/library/include/ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_convinvscale.hpp#L89) for all implementations that were instantiated. 

**Note.** This PR includes changes proposed in #1326 .

geyyer committed:
Add a scale op

geyyer committed:
Merge branch 'develop' into lwpck-1548

geyyer committed:
Merge branch 'develop' into lwpck-1548

geyyer committed:
Update the element op

geyyer committed:
Add instances

geyyer committed:
Add an example

geyyer committed:
Add a client example

geyyer committed:
Merge branch 'develop' into lwpck-1548

geyyer committed:
Add a flag check

geyyer committed:
Revert flag check addition

geyyer committed:
Fix flag check

geyyer committed:
Update d strides in example

geyyer committed:
Update d strides in client example

geyyer committed:
Merge branch 'develop' into lwpck-1548

geyyer committed:
Apply suggestions from code review

Update copyright header

Co-authored-by: Bartłomiej Kocot <barkocot@amd.com>

geyyer committed:
Move the example

geyyer committed:
Move the client example

geyyer committed:
Update element op

geyyer committed:
Update example with the new element op

geyyer committed:
Add scalar layout

geyyer committed:
Update example

geyyer committed:
Update kernel for scalar Ds

geyyer committed:
Revert kernel changes

geyyer committed:
Update element op

geyyer committed:
Update example to use scales' pointers

geyyer committed:
Format

geyyer committed:
Merge branch 'develop' into lwpck-1548

geyyer committed:
Update instances

geyyer committed:
Update client example

geyyer committed:
Move element op to unary elements

geyyer committed:
Merge branch 'develop' into lwpck-1548

andriy-ca committed:
Add CMakePresets.json

andriy-ca committed:
Merge commit '80db62f08d288f5d451753e7fc2aa4d602382892' into andriy/lwpck-1739

andriy-ca committed:
Merge commit 'a174cabfc9a5b0aa375587c141d3460c7ccd32c9' into andriy/lwpck-1739

geyyer committed:
Update element op to work with values instead of pointers

andriy-ca committed:
Merge commit '32626a6d15eaf53bb54641c15c63c2cc819d9a6e' into andriy/lwpck-1739

andriy-ca committed:
Add BF8 ConvScale Example.

andriy-ca committed:
Merge commit '6fb1f4e03fef8a80ae8b5f139b9d4750e2f1a972' into andriy/lwpck-1739

geyyer committed:
Merge branch 'develop' into lwpck-1548

geyyer committed:
Update instances to take element op as an argument

geyyer committed:
Update examples to use random scale values

andriy-ca committed:
Add instances of forward convolution with element-wise scaling for BF8 type.

andriy-ca committed:
Merge commit '0dfd142a89ef62be55ceb8d3daaa5f8b1829ee17' into andriy/lwpck-1739

andriy-ca committed:
Merge commit 'cb0645bedca3650dcaed83d599dda664c727ce38' into andriy/lwpck-1739

andriy-ca committed:
Add BF8 type to selection of DTYPES.

andriy-ca committed:
Add client_conv3d_fwd_convscale_bf8 client example.

andriy-ca committed:
Merge commit 'ac58cc5d1d4fe23fe5d08e16049d89397b8250e6' into andriy/lwpck-1739

andriy-ca committed:
Clean up CMakePresets.json

andriy-ca committed:
Apply clang-format.

andriy-ca committed:
Address reviewer's comments.

andriy-ca committed:
Merge branch 'develop' into andriy/lwpck-1739

andriy-ca committed:
Add placeholder for Grouped Convolution Forward+Elementwise Operation Profiler.

andriy-ca committed:
Merge commit 'ce66277a762e29900affbd4844d73930b78a59b5' into andriy/lwpck-1739

andriy-ca committed:
Merge branch lwpck-1739 into lwpck-1825.

andriy-ca committed:
Merge branch 'develop' into andriy/lwpck-1739

andriy-ca committed:
Update from andriy/lwpck-1739.

andriy-ca committed:
Merge remote-tracking branch 'origin/develop' into andriy/lwpck-1739

andriy-ca committed:
Merge branch 'develop' into andriy/lwpck-1739

andriy-ca committed:
WIP: Initial setup.

andriy-ca committed:
WIP: Implemented profiler interface.

andriy-ca committed:
Merge branch 'develop' into andriy/lwpck-1739

andriy-ca committed:
Merge branch 'andriy/lwpck-1739' of https://github.com/ROCm/composable_kernel into lwpck-1825

andriy-ca committed:
Merge branch 'develop' into andriy/lwpck-1739

andriy-ca committed:
ckProfiler that builds and links.

andriy-ca committed:
Merge andriy/lwpck-1739 into lwpck-1825.

andriy-ca committed:
Merge branch 'develop' into andriy/lwpck-1739

andriy-ca committed:
Merge branch 'develop' into andriy/lwpck-1739

andriy-ca committed:
Merge branch 'develop' into andriy/lwpck-1739

andriy-ca committed:
Utilities for algorithm verification.

andriy-ca committed:
Merge andriy/lwpck-1739 into andriy/lwpck-1825.

andriy-ca committed:
Merge branch 'develop' into andriy/lwpck-1739

andriy-ca committed:
Log nonstandard FP types.

andriy-ca committed:
Add tolerances for nonstandard FP types verification.

andriy-ca committed:
Merge branch 'develop' into andriy/lwpck-1739

andriy-ca committed:
Merge branch andriy/lwpck-1739 into andriy/lwpck-1825.

andriy-ca committed:
Add script for profiling grouped convolutions FWD with OUT element-wise operation.

andriy-ca committed:
Cleanup.

andriy-ca committed:
Fix formatting.

andriy-ca committed:
Fix formatting.

andriy-ca committed:
Merge branch 'develop' into andriy/lwpck-1825

andriy-ca committed:
Merge branch 'develop' into andriy/lwpck-1825

andriy-ca committed:
Merge branch 'develop' into andriy/lwpck-1825

geyyer commented on profiler/include/profiler/profile_grouped_conv_fwd_outelementop_impl.hpp:
```diff
@@ -0,0 +1,352 @@
+    {
+    case 0: break;
+    case 1:
+        input.GenerateTensorValue(GeneratorTensor_2<InDataType>{-5, 5});
```
Sometimes we have to go for smaller init values for fp8 as large values easily lead to reaching fp8 maximum of 240.0. Do these init values work fine?

andriy-ca replied:
These are the same values that are used in `profile_grouped_conv_fwd_impl.hpp`. They do result in higher errors when verified against the host convolution. Yet the errors are bounded by 16. No crashes or FP errors are reported in runtime.

What is different here, we extended ranges for `default` tensor generators. In `profile_grouped_conv_fwd_impl.hpp`, all tensor entries are generated under 1 which makes most calculations to be performed on numbers < 1.

It would be very helpful to know, which values users expect to store in the original tensors performing such convolutions.

andriy-ca replied:
I have kept input ranges to [-5,5] and reduced weights ranges to [-1, 1]. This should help addressing potential overflows on non-standard FP types.

geyyer commented on profiler/include/profiler/profile_grouped_conv_fwd_outelementop_impl.hpp:
```diff
@@ -0,0 +1,352 @@
+
+    // random scale values
+    auto scale_in = type_convert<float>(
+        type_convert<f8_t>(4.0f * float(RAND_MAX / 2 - std::rand()) / float(RAND_MAX)));
```
Is conversion to f8_t needed here?

andriy-ca replied:
Double conversion here produces `float` numbers that are representable by `FP8` without errors. It appears, scales produced this way result in fewer rounding errors.

andriy-ca replied:
In order to manage potential floating-point errors, I have also reduced the range for the scales here from [-2, 2] to [-1, 1].

geyyer commented on profiler/src/profile_grouped_conv_fwd_outelementop.cpp:
```diff
@@ -0,0 +1,222 @@
+    using ConvScale    = ck::tensor_operation::element_wise::ConvScale;
+    using ConvInvScale = ck::tensor_operation::element_wise::ConvInvscale;
+
+    // constexpr auto I1 = ck::Number<1>{};
```
Could you take care of the leftovers?

andriy-ca replied:
This has been cleaned up.

geyyer reviewed (commented):
@andriy-ca, great job, could you address the comments and this PR should be good!

andriy-ca committed:
Merge branch 'develop' into andriy/lwpck-1825

andriy-ca committed:
Reduce ranges of initial values in convolutions.

In order to keep convolution results from overflowing FP8/BF8 data types, we reduced some of the initialization ranges to [-1, 1].

geyyer reviewed (approved):
LGTM!

andriy-ca committed:
Merge branch 'develop' into andriy/lwpck-1825

andriy-ca committed:
Merge branch 'develop' into andriy/lwpck-1825
